Step1: 1단계 - 레거시 코드 리팩터링

wlghsp opened the pull request:
안녕하세요 

리뷰어님 1단계 - 레거시 코드 리팩터링 pr 올립니다.

일급 컬렉션 추가 등 리팩토링은 진행하였는데, 인스턴스 변수의 수를 3개 미만으로 줄이지 못한 점이 아쉽습니다. 

리뷰 부탁드립니다.

wlghsp committed:
chore: exception 패키지 생성 및 예외 클래스 이동

wlghsp committed:
refactor: 날짜 관련 필드 추상 클래스로 생성

wlghsp committed:
docs: 1단계 리팩토링 내역

wlghsp committed:
chore: repository 클래스 repository 패키지 생성 및 이동

wlghsp committed:
refactor: 유저 동등성 확인을 위한 equals hashcode 추가

wlghsp committed:
refactor: Answers 일급 컬렉션 생성 및 테스트 추가

wlghsp committed:
test: 답변글 작성자 확인 예외 테스트 추가

wlghsp committed:
refactor: DeleteHistory 생성일자 파라미터 제거

wlghsp committed:
refactor: QnAService 리팩토링

wlghsp committed:
test: 질문 테스트 추가

testrace commented on src/main/java/nextstep/qna/README.md:
```diff
@@ -0,0 +1,14 @@
+## 1단계 리팩토링 내역
+
+- [ ] Question 에 메시지 보내기
+  - Question 의 owner 인지 확인
+  - Answers 의 owner 인지 확인
+  - 삭제 처리
+  - DeleteHistory 생성
+- [ ] 일급컬렉션 추가 
+  - Answers
+- [ ] 인스턴스 변수 줄이기
+  - Question
+  - Answer
+- [ ] exception 패키지 생성해서 예외 클래스 이동
+- [ ] repository 패키지로 repository 이동
```
해결이 된 항목은 체크해 주시면 좋을 것 같아요.

wlghsp replied:
네 알겠습니다. :)

testrace commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -47,29 +46,30 @@ public Long getId() {
 
-    public Answer setDeleted(boolean deleted) {
-        this.deleted = deleted;
+    public Answer setDeleted() {
+        this.deleted = true;
```
setter를 활용하지 않고 메시지를 보내도록 메서드를 제공하면 어떨까요?

wlghsp replied:
네 알겠습니다.

testrace commented on src/main/java/nextstep/qna/domain/Answers.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class Answers {
+
+    private final List<Answer> answers;
+
+    private Answers(List<Answer> answers) {
+        this.answers = answers;
+    }
+
+    public static Answers createAnswers(List<Answer> answers) {
+        return new Answers(answers);
+    }
```
일급 컬렉션 활용 👍 
정적 팩토리 메서드와 생성자는 차이가 없으니 정적 팩토리 메서드는 제거해도 괜찮지 않을까요?

wlghsp replied:
네 필요한 경우에 정적 팩토리 메서드 사용해야하는데 제가 남발했던 거 같습니다. 
반영하겠습니다.

testrace commented on src/main/java/nextstep/qna/service/QnAService.java:
```diff
@@ -25,25 +24,15 @@ public class QnAService {
-                throw new CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.");
-            }
-        }
+        Question question = findQuestion(questionId);
+        question.checkIfQuestionOwner(loginUser);
+        question.checkIfAnswerOwner(loginUser);
+        question.setDeleted();
+        deleteHistoryService.saveAll(question.generateDeleteHistories());
```
깔끔하게 리팩토링 하셨네요 👍 
다만 check~ 메서드를 모두 호출을 해야한다는 사실을 모른다면 일부 조건만 충족하더라도 삭제가 가능한 상황이 됩니다.
도메인 객체가 스스로 판단해서 삭제 결과를 리턴하면 어떨까요?
```suggestion
        Question question = findQuestion(questionId);
        List<DeleteHistory> histories = question.delete(loginUser);
        deleteHistoryService.saveAll(histories);
```

wlghsp replied:
제안해주신 코드가 훨씬 깔끔하네요.
반영하겠습니다.

testrace commented on src/main/java/nextstep/qna/domain/common/BaseTime.java:
```diff
@@ -0,0 +1,10 @@
+
+import java.time.LocalDateTime;
+
+public abstract class BaseTime {
```
1단계는 인스턴스 변수의 수는 이후 단계에서 적용해 보시면 좋을 것 같아요.

wlghsp replied:
네 알겠습니다 
감사합니다.

testrace reviewed (changes requested):
안녕하세요 지호님 😃 
수강신청 미션을 함께 하게 된 송용주입니다.
리팩토링 깔끔하게 잘 해주셨네요 👍 
몇 가지 코멘트 남겨두었는데 확인해 주시고 다시 리뷰 요청해 주세요.

Commit:
refactor: 기존 matchUser 메소드 활용 유저 동등 여부 확인

Commit:
refactor: deleteQuestion 메서드 리팩토링
- Question 에 메시지 보내서 클래스 내부에서 질문 및 답변글 작성자 여부 확인
- Answers 정적 팩터리 메서드 삭제 후 생성자 사용
- Question, Answer delete 메서드 테스트 추가
- 삭제 처리 setter 가 아닌 삭제 메소드명 변경 및 반환 타입 void 로 변경

Commit:
docs: 리팩토링 실행 내역 체크

testrace commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -69,11 +68,16 @@ public void checkIfAnotherOwnerExists(NsUser loginUser) throws CannotDeleteExcep
         return "Answer [id=" + getId() + ", writer=" + writer + ", contents=" + contents + "]";
     }
+
+    public DeleteHistory delete() {
+        deleteAnswer();
+        return createDeleteHistory();
+    }
```
외부에서 이 메서드를 호출한다면 동일 작성자 여부를 검증하지 않고 삭제가 가능하게 됩니다.
1개의 답변만 삭제하더라도 checkIfAnotherOwnerExists 메서드를 항상 수행하도록 변경하면 어떨까요?

wlghsp replied:
네 알겠습니다. :)

testrace commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -50,30 +50,29 @@ public void addAnswer(Answer answer) {
     }
 
-    public Question setDeleted() {
+    public void deleteQuestion() {
```
public 으로 열려 있다보니 여전히 검증없이 삭제가 가능합니다.
private으로 변경하거나 메서드를 제거해서 항상 작성자와 일치여부를 검증하도록 하면 어떨까요?
Answer#deleteAnswer 마찬가지입니다.

wlghsp replied:
네 감사합니다.

testrace commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -88,4 +87,11 @@ private DeleteHistory createDeleteHistory() {
         return "Question [id=" + getId() + ", title=" + title + ", contents=" + contents + ", writer=" + writer + "]";
     }
+
+    public List<DeleteHistory> delete(NsUser loginUser) throws CannotDeleteException {
+        checkIfQuestionOwner(loginUser);
+        checkIfAnswerOwner(loginUser);
+        deleteQuestion();
+        return generateDeleteHistories();
```
이전의 피드백과 마찬가지로 check~ 메서드를 일일이 호출해줘야한다는걸 모른다면 작성자와 관계 없이 삭제가 가능합니다.
checkIfAnswerOwner 메서드를 직접 호출하지 않고 메시지를 보내서 결과를 받으면 어떨까요?
`List<DeleteHistory> answerDeleteHistories = answers.delete(loginUser);`

wlghsp replied:
네 '메시지를 보낸다' 명심하겠습니다.
감사합니다.

testrace commented on src/test/java/nextstep/qna/domain/QuestionTest.java:
```diff
@@ -18,24 +18,25 @@ public class QuestionTest {
     void checkIfQuestionOwnerTest() {
         assertThatThrownBy(
-                        () -> Q1.checkIfQuestionOwner(NsUserTest.SANJIGI)
+                        () -> Q1.delete(NsUserTest.SANJIGI)
```
AnswerTest에도 동일하게 delete 메서드로 예외를 검증하면 좋을 것 같아요

wlghsp replied:
네 감사합니다.

testrace reviewed (changes requested):
안녕하세요 지호님 😃 
피드백 반영 잘해주셨네요 👍 
조금만 더 개선하면 좋을 것 같아 코멘트 남겨두었습니다.
확인해 주시고 다시 리뷰 요청해 주세요.

wlghsp committed:
refactor: 삭제 메시지를 보내도록 리팩토링

- 삭제 메시지를 보내고 삭제 메서드에서 예외 발생 테스트

testrace commented on src/main/java/nextstep/qna/domain/Answer.java:
```diff
@@ -61,10 +57,11 @@ public DeleteHistory createDeleteHistory() {
     }
 
-    public void checkIfAnotherOwnerExists(NsUser loginUser) throws CannotDeleteException {
+    private void deleteAnswer(NsUser loginUser) throws CannotDeleteException {
         if (isDifferentOwner(loginUser)) {
             throw new CannotDeleteException(ANOTHER_OWNER_EXISTS_EXCEPTION_MESSAGE);
         }
+        this.deleted = true;
```
답변만 삭제하더라도 요구사항에 따라 예외를 던질 수 있게 되었네요 👍

testrace reviewed (approved):
피드백 반영 잘해주셨네요 👍 
다음 단계 진행해 주세요.
